### PR TITLE
feat: expand 'vertical splitting' to improve compaction efficiency

### PR DESCRIPTION
**Pull request by Joe-Blount**
Two key changes are made in this PR:
1: This PR adds an additional criteria for vertical splitting introduced in https://github.com/influxdata/influxdb_iox/pull/7326. This is referred to as `scenario 1` in `high_l0_overlap_split`. I considered replacing the original criteria for vertical splitting (now referred to as `scenario 2`), but we currently get better efficiency keeping both.  Since we're just starting to look at the efficiency of the split & overwrite pattern I thought I'd leave both criteria for now.  I also added another split reason so the two criteria for vertical splitting are differentiated in the simulator output.

2: changes to `DivideInitial`::`divide` to do a `min_time` based grouping of the L0 files into chains before sorting by `max_l0_created_at` and identifying `branches`.  Without this change, the vertical splits are typically undone by the subsequent compactions.  It is this grouping into "chains" that allows the benefit of the vertical splitting to realized in subsequent compactions.

Prior to posting this PR, I added tracking of the total bytes written to the simulator in https://github.com/influxdata/influxdb_iox/pull/7445 so we can better judge efficiency changes in simulator tests.

The first commit is the functional changes.  The second commit is the insta test updates.
On the insta test updates, I suggest judging efficiency changes on two criteria:
1. changes in the total bytes written (as indicated in the final output heading: `Final Output Files (xxxx written)`.  Less bytes written to fully compact is better.
2. changes in total files written (as indicated by the highest file_id in the final output).  More files means more churn of the catalog.

I'll add some comments to help understand the diff on the functional changes and comments calling out the efficiency changes in the insta test update so they're easy to spot.

helps: https://github.com/influxdata/idpe/issues/17292

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: compactor2/tests/layouts/backfill.rs [hunk 1]
```diff
@@ -1960,421 +1960,376 @@ async fn random_backfill_empty_partition() {
...
-    - "L2.?[618,805] 1.05us 100.17mb|-------------------L2.?--------------------|                                             "
-    - "L2.?[806,986] 1.05us 96.95mb                                             |-------------------L2.?-------------------| "
+    - "L2.?[772,943] 1.05us 91.75mb|--------------------------------L2.?---------------------------------|                   "
+    - "L2.?[944,986] 1.05us 23.07mb                                                                        |-----L2.?------| "
     - "Committing partition 1:"
-    - "  Soft Deleting 2 files: L1.591, L1.592"
+    - "  Soft Deleting 2 files: L1.582, L1.583"
     - "  Creating 2 files"
-    - "**** Final Output Files (4.38gb written)"
+    - "**** Final Output Files (3.86gb written)"
```
Comment (Joe-Blount) [inline compactor2/tests/layouts/backfill.rs:2325]: A modest but nice efficiency improvement: 
total file count goes from 599 -> 588 and bytes written goes from 4.38gb to 3.86gb.

File: compactor2/tests/layouts/backfill.rs [hunk 2]
```diff
@@ -6827,68 +6772,70 @@ async fn actual_case_from_catalog_1() {
     - "WARNING: file L0.161[327,333] 336ns 183.09mb exceeds soft limit 100mb by more than 50%"
     - "WARNING: file L0.162[330,338] 340ns 230.91mb exceeds soft limit 100mb by more than 50%"
     - "WARNING: file L0.163[331,338] 341ns 232.23mb exceeds soft limit 100mb by more than 50%"
-    - "**** Final Output Files (445.56gb written)"
+    - "**** Final Output Files (26.79gb written)"
```
Comment (Joe-Blount) [inline compactor2/tests/layouts/backfill.rs:6775]: A drastic efficiency improvement: 
total file count goes from 12,767 to 706, and bytes written goes from 445.56gb to 26.79gb.

This test case is what prompted the investigation leading to this PR.

**Review by Joe-Blount — COMMENTED** [review]
knob.rs has too many diffs for github to display :disappointed: 
But it has a couple efficiency improvements:
files written: 770 -> 513, bytes written: 1.29gb -> 1.02gb
files written: 779 -> 513, bytes written: 1.29gb -> 1.02gb

I didn't notice efficiency getting worse on any test case in this PR.

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs [hunk 3]
```diff
@@ -83,6 +85,57 @@ pub fn high_l0_overlap_split(
...
+                for file in &chain {
+                    let this_file_splits: Vec<i64> = split_times
+                        .iter()
+                        .filter(|split| {
+                            split >= &&file.min_time.get() && split < &&file.max_time.get()
+                        })
+                        .cloned()
+                        .collect();
+
+                    if !this_file_splits.is_empty() {
```
Comment (philjb) [inline compactor2/src/components/split_or_compact/start_level_files_to_split.rs:117]: my IDE reports this if block is identical to another if block at new line number 173.

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs [hunk 4]
```diff
@@ -193,6 +250,82 @@ pub fn select_split_times(
...
+            // This overlaps the chain, add to it.
+            if file.max_time > max_time {
+                max_time = file.max_time;
+            }
+            chain.push(file);
+        } else {
+            // file does not overlap the chain, its the start of a new chain.
+            max_time = file.max_time;
+            chains.push(chain);
+            chain = Vec::with_capacity(left);
```
Comment (philjb) [inline compactor2/src/components/split_or_compact/start_level_files_to_split.rs:282]: Is using `with_capacity` always recommended with a `Vec`?

`left` is the maximum possible length of a chain; in practice it'll be usually less. /shrug

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs [hunk 5]
```diff
@@ -193,6 +250,82 @@ pub fn select_split_times(
...
+    merged_chains
+}
+
+// get_max_l0_created_at gets the highest max_l0_created_at from all files within a vec.
+fn get_max_l0_created_at(files: Vec<ParquetFile>) -> Timestamp {
+    files
+        .into_iter()
+        .map(|f| f.max_l0_created_at)
+        .max()
+        .unwrap()
```
Comment (philjb) [inline compactor2/src/components/split_or_compact/start_level_files_to_split.rs:326]: presumes files is at least of length 1 (which is always is with the current call structure).

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs [hunk 6]
```diff
@@ -193,6 +250,82 @@ pub fn select_split_times(
...
+    chains
+}
+
+// merge_small_l0_chains takes a vector of overlapping "chains" (where a chain is vector of overlapping L0 files), and
+// attempts to merge small chains together if doing so can keep them under the given max_comapct_size.
+// This function makes no assumption about the order of the chains - if they are created by `split_into_chains`, they're
+// ordered by min_time, which is unsafe for merging L0 chains.
+pub fn merge_small_l0_chains(
+    mut chains: Vec<Vec<ParquetFile>>,
+    max_compact_size: usize,
```
Comment (philjb) [inline compactor2/src/components/split_or_compact/start_level_files_to_split.rs:297]: nit: I like units on vars. `max_compact_size_bytes`. I dunno if this is pattern breaking.

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 4 above)
Comment (Joe-Blount) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:282]: > Is using `with_capacity` always recommended with a `Vec`?

I've seen suggestions made in reviews to set the capacity when it wasn't previously.  I'm inferring we should when we know the upper bound.  

> `left` is the maximum possible length of a chain; in practice it'll be usually less. /shrug

correct.  So the question is: is it better temporarily allocate some extra memory, but do it in a single allocation, or is better to allocate the correct amount but reallocate and free potentially many times?  I think the former is preferred. so long as it isn't too much extra.  But I'd defer to @carols10cents

**Review by philjb — APPROVED** [review]
I put some nits into a branch - mostly comments and var names. nothing critical but the branch is here if you want to grab the commit. `jrb_25_improve_backlogged_test_case-pjb`

File: compactor2/src/components/split_or_compact/start_level_files_to_split.rs (hunk 6 above)
Comment (Joe-Blount) [reply compactor2/src/components/split_or_compact/start_level_files_to_split.rs:297]: I see 79 references without `_bytes` and 12 with `_bytes`.  I'd rather stick with the primary pattern for now.

**Pull request merged**